[WIP] Add support for external Prometheus mode to ScyllaDBMonitoring

czeslavo opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/scylladb/scylla-operator/blob/master/CONTRIBUTING.md)
documentation before submitting a Pull Request!
Thank you for contributing to the Scylla Operator! -->

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves https://github.com/scylladb/scylla-operator/issues/2928.

czeslavo commented on pkg/controller/scylladbmonitoring/grafana_datasource.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func makePrometheusDatasourceSpec(sm *scyllav1alpha1.ScyllaDBMonitoring) *scyllav1alpha1.GrafanaDatasourceSpec {
+	if spec := getGrafanaSpec(sm); spec != nil && len(spec.Datasources) > 0 {
+		// We only support one datasource for now.
+		ds := spec.Datasources[0]
+		if ds.Type == scyllav1alpha1.GrafanaDatasourceTypePrometheus {
+			return &ds
+		}
+	}
+	return nil
```
There are two valid cases here:
1. There is 1 datasource with Prometheus type defined -> return the datasource.
2. There's no such datasource defined -> return nil.

I intended not to duplicate the validation logic here, but instead rely on it.

czeslavo commented on pkg/controller/scylladbmonitoring/sync_prometheus.go:
```diff
@@ -584,3 +576,63 @@ func (smc *Controller) syncPrometheus(
+
+		resources.Prometheus, _, err = makePrometheus(sm, soc)
+		renderErrors = append(renderErrors, err)
+	}
```
Converted into a switch.

czeslavo commented on pkg/controller/scylladbmonitoring/sync_grafana.go:
```diff
@@ -336,6 +345,7 @@ func (smc *Controller) syncGrafana(
 
 	var requiredDeployment *appsv1.Deployment
 	// Trigger restart for inputs that are not live reloaded.
+	// TODO: add custom TLS certs and auth token Secrets for Prometheus datasource.
```
Removed it.

czeslavo replied:
It's been handled in https://github.com/scylladb/scylla-operator/pull/2943/commits/ca10f43bc8dd851a43da5a823b79a1ac886c2e66.

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring_webhooks.go:
```diff
@@ -0,0 +1,76 @@
+	f := framework.NewFramework("scylladbmonitoring")
+
+	g.It("should forbid invalid requests", func() {
+		ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+		defer cancel()
```
nit: use `g.SpecContext` and set timeout on the ginkgo node
nit: I'd also remove the test timeout altogether, it's limited by the suite anyway

czeslavo commented on pkg/controller/scylladbmonitoring/sync_grafana_test.go:
```diff
@@ -723,3 +971,278 @@ spec:
+`, "\n"),
+		},
+		{
+			name: "custom prometheus datasource without TLS",
```
Fixed.

czeslavo commented on pkg/controller/scylladbmonitoring/sync_grafana_test.go:
```diff
@@ -723,3 +971,345 @@ spec:
+			t.Parallel()
+			_, objString, err := makeGrafanaProvisionings(tc.sm)
+			if err != nil {
+				t.Fatalf("unexpected error: %v\nRendered object:\n%s", err, objString)
```
That's a good point, no need for that.

czeslavo commented on pkg/api/scylla/validation/scylladbmonitoring_validation.go:
```diff
@@ -0,0 +1,256 @@
+	if !oslices.ContainsItem(allowedDatasourceNames, ds.Name) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("name"), ds.Name, allowedDatasourceNames))
+	}
+
```
Moved from Prometheus-specific function.

czeslavo commented on pkg/api/scylla/validation/scylladbmonitoring_validation.go:
```diff
@@ -0,0 +1,256 @@
+
+	if ds.URL == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("url"), "must be specified for Prometheus datasource"))
+	}
```
Moved.

czeslavo commented on pkg/controller/scylladbmonitoring/conditions.go:
```diff
@@ -1,8 +1,10 @@
+	grafanaControllerProgressingCondition          = "GrafanaControllerProgressing"
+	grafanaControllerDegradedCondition             = "GrafanaControllerDegraded"
+	grafanaControllerConfigMapReferenceProgressing = "GrafanaControllerConfigMapReferenceProgressing"
+	grafanaControllerSecretReferenceProgressing    = "GrafanaControllerSecretReferenceProgressing"
```
I wasn't sure what the convention was in such cases. I noticed it's aggregated, but it felt weird to me to set the same condition multiple times. Removed.

rzetelskik replied:
The aggregating function ignores the types altogether, so it just seems like unnecessary noise to me - alternatively we could have some private, internal types, but I wouldn't expose them in `conditions.go`.

rzetelskik commented on pkg/api/scylla/validation/scylladbmonitoring_validation.go:
```diff
@@ -0,0 +1,256 @@
+	if ds.Type == scyllav1alpha1.GrafanaDatasourceTypePrometheus {
+		allErrs = append(allErrs, validateScyllaDBMonitoringGrafanaPrometheusDatasource(&ds, fldPath)...)
+	}
+
```
IMO cross-checking improves UX as we're being verbose about the options which won't be used. This comment surprises me a lot because we implemented similar cross-checking recently with ScyllaDBManagerTask and it doesn't seem like that big of a mess to me: https://github.com/scylladb/scylla-operator/blob/daaccfb82aaa2e10ee0272acdfc0e7bfe3a6068f/pkg/api/scylla/validation/scylladbmanagertask_validation.go#L189-L225? For every TypeX, you essentially do:
- if TypeX, then validate XOptions
- if not TypeX, then validate XOptions not set

czeslavo replied:
Addressed. Posted https://github.com/scylladb/scylla-operator/pull/2961.

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring_webhooks.go:
```diff
@@ -0,0 +1,102 @@
+			o.Expect(err).To(e.expectedErrMatcherFunc(smCopy))
+		}, g.NodeTimeout(testTimeout))
+	},
+		g.Entry("with valid spec", &entry{
```
```suggestion
		g.Entry("with acceptance when a valid ScyllaDBMonitoring", &entry{
```
otherwise this will read as `ScyllaDBMonitoring webhook should respond with valid spec [It] is created`
as opposed to `ScyllaDBMonitoring webhook should respond with acceptance when a valid ScyllaDBMonitoring [It] is created`

czeslavo replied:
Fixed.

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring_webhooks.go:
```diff
@@ -0,0 +1,102 @@
+			modifierFuncs:          nil,
+			expectedErrMatcherFunc: func(sm *scyllav1alpha1.ScyllaDBMonitoring) o.OmegaMatcher { return o.Succeed() },
+		}),
+		g.Entry("with invalid spec", &entry{
```
ditto

czeslavo replied:
Fixed.

czeslavo commented on pkg/api/scylla/validation/scylladbmonitoring_validation.go:
```diff
@@ -0,0 +1,267 @@
+
+	if ds.Type == scyllav1alpha1.GrafanaDatasourceTypePrometheus {
+		allErrs = append(allErrs, validateScyllaDBMonitoringGrafanaPrometheusDatasource(&ds, fldPath)...)
+	}
```
I left it separate as my idea for `validateScyllaDBMonitoringGrafanaDatasourceOptions` was to only include these cross-validation rules. But after looking at it again, it makes more sense to put it all together. Updated.

czeslavo commented:
/retest

czeslavo commented:
/test images
/retest

czeslavo commented:
/retest

czeslavo commented:
/retest

czeslavo commented:
/retest

czeslavo commented:
Rebased onto master and added E2E tests to this PR, as we're blocked from merging it anyway (the initial intention was to keep tests separate to streamline the review process).

Programmatically skipping the OpenShift-specific test case due to the lack of an OpenShift-only suite (https://github.com/scylladb/scylla-operator/pull/2972, which is also blocked from merging due to the unstable multi-DC suite).

/hold for https://github.com/scylladb/scylla-operator/pull/2972

czeslavo commented:
/retest

czeslavo commented:
/test e2e-openshift-aws-parallel

czeslavo commented:
/hold cancel

czeslavo commented:
/test e2e-openshift-aws-parallel

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:
```diff
@@ -188,12 +244,12 @@ func awaitScyllaDBMonitoringRollout(ctx context.Context, f *framework.Framework,
-	//  - prometheus configuration is asynchronous without any acknowledgement
-	//  - grafana configuration is asynchronous without any acknowledgement
-	// Some of these may be fixable by manually verifying it in the operator sync loop so it can also be
-	// consumed by clients, but it's a bigger effort.
```
Why is this being removed? I don't see any code changes that would explain it (it's still polling), but I might have lost some context.

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:
```diff
@@ -436,3 +497,357 @@ func decodeGrafanaDashboardFromGZBase64String(s string) (*grafanaDashboard, erro
+	framework.By("Creating a Prometheus instance")
+	prom, err := f.PrometheusOperatorAdminClient().MonitoringV1().Prometheuses(f.Namespace()).Create(ctx, &monitoringv1.Prometheus{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      prometheusNameForScyllaDBMonitoring(sm),
```
nit: I found it very confusing how the external prometheus instance was wired into ScyllaDBMonitoring, i.e. that is used the same naming convention as the managed instance. Should it take a name (service name??) as parameter?

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:
```diff
@@ -436,3 +497,357 @@ func decodeGrafanaDashboardFromGZBase64String(s string) (*grafanaDashboard, erro
+				},
+				ServiceMonitorSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"scylla-operator.scylladb.com/scylladbmonitoring-name": sm.Name,
```
same here, I think the wiring would be much more understandable if the constructors were "scylladbmonitoring agnostic" - i.e. they took everything needed for wiring with SDBM as parameters, instead of creating it from ScyllaDBMonitoring the same way we do for managed instances

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:
```diff
@@ -436,3 +497,357 @@ func decodeGrafanaDashboardFromGZBase64String(s string) (*grafanaDashboard, erro
+	g.GinkgoHelper()
+
+	framework.By("Creating a Prometheus instance with TLS")
+	tlsSecretName := prometheusTLSSecretNameForScyllaDBMonitoring(sm)
```
Again, lost a bit of context. Who creates this? Are we wiring the external prometheus instance with certs generated by the operator?

rzetelskik commented on test/e2e/set/scylladbmonitoring/scylladbmonitoring.go:
```diff
@@ -171,6 +174,41 @@ var _ = g.Describe("ScyllaDBMonitoring", func() {
+							TLS: &scyllav1alpha1.GrafanaDatasourceTLSSpec{
+								InsecureSkipVerify: false,
+								CACertConfigMapRef: &scyllav1alpha1.LocalObjectKeySelector{
+									Name: "openshift-service-ca.crt",
```
Q: how is this created in the namespace?

rzetelskik reviewed (commented):
Solid :muscle: 
left some minor comments

czeslavo commented:
/test images
/test olm-bundle-openshift-preflight
